drivers/hwinfo/hwinfo_sam.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2019 Aurelien Jarno
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <hwinfo.h>
+#include <init.h>
+#include <soc.h>
+#include <string.h>
+
+static u8_t sam_uid[16];
+
+ssize_t _impl_hwinfo_get_device_id(u8_t *buffer, size_t length)
+{
+	if (length > sizeof(sam_uid)) {
+		length = sizeof(sam_uid);
+	}
+
+	memcpy(buffer, sam_uid, length);
+
+	return length;
+}
+
+/* On the Atmel SAM SoC series, the device id is located in the flash
+ * controller. The controller can either present the flash area containing
+ * the code, the unique identifier or the user signature area at the flash
+ * location. Therefore the function reading the device id must be executed
+ * from RAM with the interrupts disabled. To avoid executing this complex
+ * code each time the device id is requested, we do this at boot time at save
+ * the 128-bit value into RAM.
+ */
+__ramfunc static int hwinfo_sam_init(struct device *arg)
+{
+	Efc *efc = (Efc *)DT_FLASH_DEV_BASE_ADDRESS;
+	u8_t *flash = (u8_t *)CONFIG_FLASH_BASE_ADDRESS;
+	u32_t fmr;
+	int key;
+
+	/* Disable interrupts. */
+	key = irq_lock();
+
+	/* Disable code loop optimization and sequential code optimization. */
+	fmr = efc->EEFC_FMR;
+	efc->EEFC_FMR = (fmr & (~EEFC_FMR_CLOE)) | EEFC_FMR_SCOD;
+
+	/* Switch the flash controller to the unique identifier area. The flash
+	 * is not available anymore, hence we have to wait for it to be *NOT*
+	 * ready.
+	 */
+	efc->EEFC_FCR = EEFC_FCR_FKEY_PASSWD | EEFC_FCR_FCMD_STUI;
+	while ((efc->EEFC_FSR & EEFC_FSR_FRDY) == EEFC_FSR_FRDY) {
+		/* Wait */
```

Comment:
Is there a way to protect against this busy loop? E.g. what if efc does not become not-ready? Also, just wondering: what kind of latency do we expect here?

Reply:
The latency is very small, ie 1 flash cycle, which is on the AHB bus. By default it runs at half the CPU speed, so the latency is at maximum 2 CPU cycles. The while loop is the recommended way from the datasheet, and is mostly there as a kind of barrier as efc is in a strongly ordered region. I have tried to put a counter in the while loop, it never gets incremented. Therefore I am not sure we need to protect against a busy loop.

Reply:
I see. From functional safety perspective, having busy loops at boot time is a bit scary - I wonder if we at least have a watchdog enabled at this point.

It looks like this is post-kernel initialization, so some mechanisms are already available to protect from busy loops, no?

Reply:
A watchdog is maybe a bit over engineered. A counter would also do the job ;-)